src/liboslexec/string_hash.h

```diff
@@ -0,0 +1,667 @@
+// Copyright Contributors to the Open Shading Language project.
+// SPDX-License-Identifier: BSD-3-Clause
+// https://github.com/imageworks/OpenShadingLanguage
+
+#pragma once
+
+#include <stdint.h>
+#include <utility>
+
+//#include "oslexec_pvt.h"
+
+#ifdef __CUDA_ARCH__
+#include <optix.h>
+
+OSL_NAMESPACE_ENTER
+
+// Create an OptiX variable for each of the 'standard' strings declared in
+// <OSL/strdecls.h>.
+
+namespace DeviceStrings {
+
+#define STATIC_INLINE __device__ __forceinline__ constexpr
+
+// BLATANTLY COPIED FROM ${OIIO}/src/include/OpenImageIO/hash.h
+//
+#if defined(FARMHASH_UINT128_T_DEFINED)
+STATIC_INLINE uint64_t Uint128Low64(const uint128_t x) {
+  return static_cast<uint64_t>(x);
+}
+STATIC_INLINE uint64_t Uint128High64(const uint128_t x) {
+  return static_cast<uint64_t>(x >> 64);
+}
+STATIC_INLINE uint128_t Uint128(uint64_t lo, uint64_t hi) {
+  return lo + (((uint128_t)hi) << 64);
+}
+STATIC_INLINE void CopyUint128(uint128_t &dst, const uint128_t src) { dst = src; }
+#else
+typedef std::pair<uint64_t, uint64_t> uint128_t;
+STATIC_INLINE uint64_t Uint128Low64(const uint128_t x) { return x.first; }
+STATIC_INLINE uint64_t Uint128High64(const uint128_t x) { return x.second; }
+STATIC_INLINE uint128_t Uint128(uint64_t lo, uint64_t hi) { return uint128_t(lo, hi); }
+STATIC_INLINE void CopyUint128(uint128_t &dst, const uint128_t src) { dst.first  = src.first;
+                                                                      dst.second = src.second; }
+#endif
+
+// This is intended to be a reasonably good hash function.
+// May change from time to time, may differ on different platforms, may differ
+// depending on NDEBUG.
+STATIC_INLINE uint64_t Hash128to64(uint128_t x) {
+  // Murmur-inspired hashing.
+  const uint64_t kMul = 0x9ddfea08eb382d69ULL;
+  uint64_t a = (Uint128Low64(x) ^ Uint128High64(x)) * kMul;
+  a ^= (a >> 47);
+  uint64_t b = (Uint128High64(x) ^ a) * kMul;
+  b ^= (b >> 47);
+  b *= kMul;
+  return b;
+}
```

Comment:
Should we eventually migrate these to the OIIO side, i.e. adorn hash.h with the proper decorators for Cuda for the things we wish to use?

Reply:
Already been looking into that.